ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -155,87 +166,94 @@
               logger.info("Stopped listening to cross colo ZK server {}", zkConnectStr);
             }
           } catch (Exception e) {
-            initializationException.compareAndSet(null, e);
+            setException(isLocalDc, e);
           } finally {
             initializationAttemptComplete.countDown();
           }
         }, false).start();
       }
-      initializationAttemptComplete.await();
-    } catch (Exception e) {
-      initializationException.compareAndSet(null, e);
+      try {
+        initializationAttemptComplete.await();
+      } catch (InterruptedException e) {
+        initializationException.compareAndSet(null, e);
+      }
     }
-    if (initializationException.get() == null) {
+    if (initializationException.get() != null) {
+      helixClusterManagerMetrics.initializeInstantiationMetric(false, remoteInitializationFailedCount.get());
+      close();
+      throw new IOException("Encountered exception while parsing json, connecting or initializing in the local DC",
+          initializationException.get());
+    } else {
       // resolve the status of all partitions before completing initialization
       for (AmbryPartition partition : partitionMap.values()) {
         partition.resolvePartitionState();
       }
       initializeCapacityStats();
-      helixClusterManagerMetrics.initializeInstantiationMetric(true);
+      helixClusterManagerMetrics.initializeInstantiationMetric(true, remoteInitializationFailedCount.get());
       helixClusterManagerMetrics.initializeXidMetric(currentXid);
       helixClusterManagerMetrics.initializeDatacenterMetrics();
       helixClusterManagerMetrics.initializeDataNodeMetrics();
       helixClusterManagerMetrics.initializeDiskMetrics();
       helixClusterManagerMetrics.initializePartitionMetrics();
       helixClusterManagerMetrics.initializeCapacityMetrics();
-    } else {
-      helixClusterManagerMetrics.initializeInstantiationMetric(false);
-      close();
-      throw new IOException("Encountered exception while parsing json, connecting to Helix or initializing",
-          initializationException.get());
     }
     localDatacenterId = dcToDcZkInfo.get(clusterMapConfig.clusterMapDatacenterName).dcZkInfo.getDcId();
     partitionSelectionHelper =
         new PartitionSelectionHelper(partitionMap.values(), clusterMapConfig.clusterMapDatacenterName);
   }
 
+  private void setException(boolean isLocalDc, Exception e) {
+    if (isLocalDc) {
+      initializationException.compareAndSet(null, e);
+    } else {
+      remoteInitializationException.compareAndSet(null, e);
+      remoteInitializationFailedCount.incrementAndGet();
```

Comment:
I am thinking in a very rare scenario, for a single remote datacenter, the initialization failure could be counted twice. The first time in `ClusterChangeHandler`, the second time may be in the initialization thread after `ClusterChangeHandler` is added as listener.

Reply:
True. Probably not a big deal, but I have changed this to use a hash map to make it more accurate.